// mx-database/lib/Util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Util = void 0;
const promise_1 = require("./promise");
// 拷贝函数 是否包括函数拷贝
function func_copy(obj, bFunc = false) {
    var out;
    if (obj instanceof Array) {
        out = [];
        // 数组这里就
        for (let i = 0; i < obj.length; i++) {
            out[i] = func_copy(obj[i], bFunc);
        }
    }
    else if (obj instanceof Buffer) {
        // buffer 这里就不拷贝算了
        out = Buffer.from(obj);
    }
    else if (typeof obj == 'object') {
        out = {};
        for (var key in obj) {
            if (key == 'clone' || key == 'global') {
                continue;
            }
            if (typeof obj[key] == 'function' && !bFunc) {
                continue;
            }
            if (obj[key] == null) {
                out[key] = null;
            }
            else
                out[key] = func_copy(obj[key], false);
        }
    }
    else {
        out = obj;
    }
    return out;
}
function func_attach(dst, src) {
    if (!dst || !src)
        return;
    for (var key in src) {
        dst[key] = exports.Util.copy(src[key]);
    }
    return dst;
}
function promise_setInterval(...args) {
    let [name, callback, time] = args;
    if (typeof name != 'string') {
        name = '';
        callback = args[0];
        time = args[1];
    }
    let handle;
    function update() {
        promise_1.createPromise(callback()).finally(function () {
            if (handle) {
                if (handle.refresh) {
                    handle.refresh();
                }
                else {
                    handle = setTimeout(update, time);
                }
            }
        });
    }
    handle = setTimeout(update, time);
    handle['name'] = name;
    return {
        stop: function () {
            clearTimeout(handle);
        }
    };
}
exports.Util = {
    copy: func_copy,
    attach: func_attach,
    setInterval: promise_setInterval
};
